### ISSUE TITLE
Prefer List.of() over List.from() when input and output types are the same

### ISSUE THREAD

**Issue by jamesderlin** [dart-lang/linter#2066]
AFAIK the difference between `List.of()` and `List.from()` (and likewise, `Map.of()`/`Map.from()`, `Set.of()`/`Set.from()`, and various other collection types) is that `.of()` takes an argument of the same type as what it returns and enforces it at compilation time, and that `.from()` allows potentially unsafe downcasting and enforces convertibility at runtime.

This distinction is subtle (and possibly confusing), and I've seen people use `.from()` unnecessarily, which presumably incurs some additional runtime cost.  It'd be nice if there were a lint that recommended using `.of()` when both the input and output types are the same.

**Examples**
```dart
var intList = [1, 2, 3];

var copy = List<int>.from(intList); // Lint. Suggest List<int>.of() instead.
var numList = List<num>.from(intList); // OK
Ideally, provide motivating examples w/ rationale.

intList = List<int>.from(numList); // OK

var unspecifedList = List.from(intList); // OK; can be flagged by [implicit_dynamic_type]
```

**Comment by jacob314** [comment]
It would be good to prioritize fixing this issue. A quick search suggests there are 1000s of violations we could fix in g3 and running dart-code-metrics on the devtools package verified there were lots of easily fixable issues in devtools.
